### PR TITLE
Added legend

### PR DESCRIPTION
**Pull request by lorem--ipsum**

### CODE REVIEW

File: src/LineChart.ts [hunk 1]
```diff
@@ -35,11 +35,15 @@ module n3Charts {
...
         ['container', Factory.Container, element[0]],
         ['tooltip', Factory.Tooltip, element[0]],
+        ['legend', Factory.Legend, element[0]],
         ['transitions', Factory.Transition],
         ['x-axis', Factory.Axis, Factory.Axis.SIDE_X],
         ['y-axis', Factory.Axis, Factory.Axis.SIDE_Y],
-        ['series-column', Factory.Series.Column],
+
+        // This order is important, otherwise it can mess up with the tooltip
+        // (and you don't want to mess up with a tooltip, trust me).
```
Comment (chaosmail) [inline src/LineChart.ts]: I am wondering why is this? I moved columns as the first series, to have this specific order with multiple types "Column, Area, Line, Dot"

File: src/LineChart.ts [hunk 2]
```diff
@@ -70,6 +74,12 @@ module n3Charts {
         eventMgr.on('out.directive', scope.onDatumOut);
       });
 
+      eventMgr.on('legend-click.directive', (series) => {
```
Comment (chaosmail) [inline src/LineChart.ts]: Can we find a better name for this event namespace `legend-click.directive`?

File: src/factories/Legend.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
+module n3Charts.Factory {
+  'use strict';
+
+  export class Legend extends Utils.BaseFactory {
+
+    // We might wanna find another name for this : it's not an SVG (same as for
+    // the tooltip).
+    // [lorem--ipsum]
+    private svg:D3.Selection;
```
Comment (chaosmail) [inline src/factories/Legend.ts]: alright, we can call it `div`

File: src/factories/Legend.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    constructor(private element: HTMLElement) {
+      super();
+    }
+
+    create() {
+      this.createLegend();
+    }
+
+    createLegend() {
+      var svg = this.svg = d3.select(this.element)
```
Comment (chaosmail) [inline src/factories/Legend.ts]: `var svg = ` can be removed

File: src/factories/Legend.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        s.select('.icon').style('background-color', (d) => d.color);
+        s.select('.label').text((d) => d.label);
+      };
+
+      var legendItems = this.svg.selectAll('.item')
+        .data(options.series);
+
+      legendItems.enter().call(init);
+      legendItems.call(update);
+      legendItems.exit().remove();
```
Comment (chaosmail) [inline src/factories/Legend.ts]: I see what you did there, and I like it :)

File: src/utils/EventManager.ts [hunk 6]
```diff
@@ -12,6 +12,9 @@ module n3Charts.Utils {
       'over',   // on mouse over a data point or column
       'out',   // on mouse out of a data point or column
       'click',   // on click on a data point or column
+      'legend-click',   // on click on a legend item
+      'legend-over',   // on mouse over on a legend item
+      'legend-out',   // on mouse out on a legend item
```
Comment (chaosmail) [inline src/utils/EventManager.ts]: we should use this names also for the other events, like `data-click`, what do you think?

File: src/factories/Legend.ts [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    update(data:Utils.Data, options:Utils.Options) {
+      // Get the container dimensions
+      var container = <Factory.Container> this.factoryMgr.get('container');
+      var dim: IDimension = container.getDimensions();
+
+      var init = (s) => {
+        var items = s.append('div').attr({'class': 'item'})
+          .call(this.eventMgr.legendClick());
```
Comment (chaosmail) [inline src/factories/Legend.ts]: `legendClick` should be defined in this class, not in the EventManager. This makes it easier to keep track of things that belong together

File: src/utils/EventManager.ts [hunk 8]
```diff
@@ -59,5 +63,13 @@ module n3Charts.Utils {
         });
       };
     }
+
+    legendClick() {
```
Comment (chaosmail) [inline src/utils/EventManager.ts]: Can we move this to the Legend class?

File: test/spec/LineChart.spec.ts [hunk 9]
```diff
@@ -1,6 +1,6 @@
   /// <reference path='test.spec.ts' />
 
-describe('n3Charts.Utils.BaseFactory', () => {
+describe('LineChart directive', () => {
```
Comment (chaosmail) [inline test/spec/LineChart.spec.ts]: can we call this `n3Charts.LineChart`?

**Comment by chaosmail** [comment]
awesome, great PR; there are only small remarks mostly about renaming stuff. Including sass is great also, but is there are reason the files all start with an underscore, like `_axes.scss`?
Can't wait to merge this and play around :)

File: src/LineChart.ts (hunk 1 above)
Comment (lorem--ipsum) [inline src/LineChart.ts]: Well if a column happens to be behind an area, the tooltips events won't fire.

File: src/LineChart.ts (hunk 2 above)
Comment (lorem--ipsum) [inline src/LineChart.ts]: Hmmmm I kind of fell short on finding names for these events :-)
Maybe we can delay that while this is still an unreleased beta ? When we have more events, we'll figure out a better naming convention ? Alternatively, if you have ideas, fell free to make some changes.

File: src/factories/Legend.ts (hunk 3 above)
Comment (lorem--ipsum) [inline src/factories/Legend.ts]: Hmmm it doesn't _have_ to be a div either... Too bad  `element` is already taken...

File: src/factories/Legend.ts (hunk 4 above)
Comment (lorem--ipsum) [inline src/factories/Legend.ts]: :+1:

File: src/utils/EventManager.ts (hunk 6 above)
Comment (lorem--ipsum) [inline src/utils/EventManager.ts]: Same as my other comment on events names, I think we need to wait until we see a broader picture for the events...

File: src/factories/Legend.ts (hunk 7 above)
Comment (lorem--ipsum) [inline src/factories/Legend.ts]: Yup, very true.

File: src/utils/EventManager.ts (hunk 8 above)
Comment (lorem--ipsum) [inline src/utils/EventManager.ts]: Will do !

File: test/spec/LineChart.spec.ts (hunk 9 above)
Comment (lorem--ipsum) [inline test/spec/LineChart.spec.ts]: Sure !

**Comment by lorem--ipsum** [comment]
The is a convention when using Sass : files that start with an underscore are considered by the transpiler  as non primary files, which mean they're transpiled only if they're 'required' somewhere. It allows the transpiler to safely ignore some files.

File: src/LineChart.ts (hunk 1 above)
Comment (chaosmail) [inline src/LineChart.ts]: alright, this makes sense - maybe we could add this explanation into the comment

**Comment by chaosmail** [comment]
nice, everything looks good :)

**Comment by satheesh786** [comment]
I want to run angular material through Dreamviewer IDE software,but its not working..can any one help me..

Thank you advance.

**Comment by chaosmail** [comment]
Hi @satheesh786 

I am sorry, this question is not at all related either with this project nor with this PR. Please try the following links to find help/solutions/support: [http://bfy.tw/qJu](http://bfy.tw/qJu) or here [https://github.com/angular/material](https://github.com/angular/material)

Best,
Christoph

**Comment by satheesh786** [comment]
got it ..thank you

**Pull request merged**